File: Core/engine/lib/src/providers/discover.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../rust/data.dart';
import '../rust/dto.dart';
import '../rust/dto_social.dart';
import 'pagination.dart';

/// Одна курсорная страница из ядра: элементы + курсор на следующую (`null` — конец).
class CursorSlice<T> {
  final List<T> items;
  final String? nextCursor;

  const CursorSlice({required this.items, this.nextCursor});
}

/// База для курсорной пагинации поверх моста, с фильтром [F].
///
/// Состояние — [PagedList] (общая форма для UI): `items` накапливаются,
/// `hasMore = nextCursor != null`, `nextOffset` тут просто длина списка.
/// Курсор живёт в [_cursor], фильтр — в [filter]; смена фильтра через
/// [setFilter] сбрасывает курсор и перечитывает с нуля.
abstract class CursorPagedNotifier<T, F> extends AsyncNotifier<PagedList<T>> {
  int get pageSize => 80;

  String? _cursor;
  late F _filter;

  /// Текущий фильтр.
  F get filter => _filter;

  /// Фильтр по умолчанию для первой загрузки.
  F initialFilter();

  /// Загрузить страницу: с [cursor] (хвост) или с нуля (`cursor == null`).
  Future<CursorSlice<T>> fetch({required F filter, String? cursor});

  @override
  Future<PagedList<T>> build() async {
    _filter = initialFilter();
    _cursor = null;
    final slice = await fetch(filter: _filter, cursor: null);
    _cursor = slice.nextCursor;
    return PagedList<T>(
      items: slice.items,
      hasMore: slice.nextCursor != null,
      nextOffset: slice.items.length,
    );
  }

  /// Догрузить следующую страницу по несомому курсору.
  Future<void> loadMore() async {
    final current = state.value;
    if (current == null ||
        !current.hasMore ||
        current.loadingMore ||
        _cursor == null) {
      return;
    }

    state = AsyncData(current.copyWith(loadingMore: true));
    try {
      final slice = await fetch(filter: _filter, cursor: _cursor);
      _cursor = slice.nextCursor;
      state = AsyncData(
        current.copyWith(
          items: [...current.items, ...slice.items],
          hasMore: slice.nextCursor != null,
          loadingMore: false,
          nextOffset: current.items.length + slice.items.length,
        ),
      );
    } catch (_) {
      // Хвост упал — оставляем показанное, снимаем флаг догрузки.
      state = AsyncData(current.copyWith(loadingMore: false));
    }
  }

  /// Перечитать с нуля по текущему фильтру.
  Future<void> refresh() async {
    _cursor = null;
    state = const AsyncLoading();
    state = await AsyncValue.guard(() async {
      final slice = await fetch(filter: _filter, cursor: null);
      _cursor = slice.nextCursor;
      return PagedList<T>(
        items: slice.items,
        hasMore: slice.nextCursor != null,
        nextOffset: slice.items.length,
      );
    });
  }

  /// Применить новый фильтр: сбросить курсор и перечитать с нуля.
  void setFilter(F next) {
    _filter = next;
    refresh();
  }
}

/// Счётчики каталога (`discover/summary`).
final discoverSummaryProvider =
    FutureProvider.autoDispose<DiscoverSummaryDto>((ref) {
  return discoverSummary();
});

/// Теги-жанры каталога (`discover/tags`).
final discoverTagsProvider = FutureProvider.autoDispose<List<TagDto>>((ref) {
  return discoverTags();
});

/// «В центре внимания» (`discover/spotlight`): курируемые карточки (артист|альбом),
/// без пагинации. Лента горизонтального скролла под FeaturedHero.
final discoverSpotlightProvider =
    FutureProvider.autoDispose<SpotlightFeedDto>((ref) {
  return discoverSpotlight();
});

/// Случайный id из каталога (`discover/random`); `kind` — album/artist/track.
/// Возвращает `null`, если каталог по фильтру пуст.
final discoverRandomProvider =
    FutureProvider.autoDispose.family<String?, String?>((ref, kind) {
  return discoverRandom(kind: kind);
});

/// Альбомы, сгруппированные по году выпуска (`discover/albums_by_year`).
/// `(years, perYear, kind)` — сколько лет назад, сколько на год, тип релиза.
final discoverAlbumsByYearProvider = FutureProvider.autoDispose
    .family<List<AlbumYearBucketDto>, AlbumsByYearArgs>((ref, args) {
  return discoverAlbumsByYear(
    years: args.years,
    perYear: args.perYear,
    kind: args.kind,
  );
});

class AlbumsByYearArgs {
  final int years;
  final int perYear;
  final String? kind;

  const AlbumsByYearArgs({
    required this.years,
    required this.perYear,
    this.kind,
  });

  @override
  int get hashCode => Object.hash(years, perYear, kind);

  @override
  bool operator ==(Object other) =>
      other is AlbumsByYearArgs &&
      years == other.years &&
      perYear == other.perYear &&
      kind == other.kind;
}

/// Каталог артистов: курсорная подгрузка + фильтры (`sort`/`tag`/`q`).
/// Состояние — [PagedList] (как у offset-списков), но хвост двигает курсор:
/// `hasMore = nextCursor != null`.
final discoverArtistsProvider =
    AsyncNotifierProvider<DiscoverArtistsNotifier, PagedList<ArtistCardDto>>(
  DiscoverArtistsNotifier.new,
);

class DiscoverArtistsNotifier
    extends CursorPagedNotifier<ArtistCardDto, ArtistFilter> {
  @override
  ArtistFilter initialFilter() => const ArtistFilter();

  @override
  Future<CursorSlice<ArtistCardDto>> fetch({
    required ArtistFilter filter,
    String? cursor,
  }) async {
    final page = await discoverArtists(
      limit: pageSize,
      cursor: cursor,
      sort: filter.sort,
      tag: filter.tag,
      q: filter.q,
    );
    return CursorSlice(items: page.items, nextCursor: page.nextCursor);
  }

  /// Сменить сортировку (`popular`/`trending`/…); сбрасывает и перезагружает.
  void setSort(String? sort) => setFilter(filter.copyWith(sort: () => sort));

  /// Сменить тег-фильтр; сбрасывает и перезагружает.
  void setTag(String? tag) => setFilter(filter.copyWith(tag: () => tag));

  /// Сменить поисковый запрос; сбрасывает и перезагружает.
  void setQuery(String? q) => setFilter(filter.copyWith(q: () => q));
}

class ArtistFilter {
  final String? sort;
  final String? tag;
  final String? q;

  const ArtistFilter({this.sort, this.tag, this.q});

  ArtistFilter copyWith({
    String? Function()? sort,
    String? Function()? tag,
    String? Function()? q,
  }) {
    return ArtistFilter(
      sort: sort != null ? sort() : this.sort,
      tag: tag != null ? tag() : this.tag,
      q: q != null ? q() : this.q,
    );
  }
}

/// Каталог альбомов: курсорная подгрузка + фильтры (`sort`/`kind`/`q`).
final discoverAlbumsProvider =
    AsyncNotifierProvider<DiscoverAlbumsNotifier, PagedList<AlbumCardDto>>(
  DiscoverAlbumsNotifier.new,
);

class DiscoverAlbumsNotifier
    extends CursorPagedNotifier<AlbumCardDto, AlbumFilter> {
  @override
  AlbumFilter initialFilter() => const AlbumFilter();

  @override
  Future<CursorSlice<AlbumCardDto>> fetch({
    required AlbumFilter filter,
    String? cursor,
  }) async {
    final page = await discoverAlbums(
      limit: pageSize,
      cursor: cursor,
      sort: filter.sort,
      kind: filter.kind,
      q: filter.q,
    );
    return CursorSlice(items: page.items, nextCursor: page.nextCursor);
  }

  /// Сменить сортировку; сбрасывает и перезагружает.
  void setSort(String? sort) => setFilter(filter.copyWith(sort: () => sort));

  /// Сменить тип релиза (`album`/`ep`/…); сбрасывает и перезагружает.
  void setKind(String? kind) => setFilter(filter.copyWith(kind: () => kind));

  /// Сменить поисковый запрос; сбрасывает и перезагружает.
  void setQuery(String? q) => setFilter(filter.copyWith(q: () => q));
}

class AlbumFilter {
  final String? sort;
  final String? kind;
  final String? q;

  const AlbumFilter({this.sort, this.kind, this.q});

  AlbumFilter copyWith({
    String? Function()? sort,
    String? Function()? kind,
    String? Function()? q,
  }) {
    return AlbumFilter(
      sort: sort != null ? sort() : this.sort,
      kind: kind != null ? kind() : this.kind,
      q: q != null ? q() : this.q,
    );
  }
}

File: Core/shared/crates/sc-bff/src/endpoints/discover.rs
use sc_domain::{
    AlbumCard, AlbumYearBucket, ArtistCard, CursorPage, DiscoverSummary, SpotlightItem, Tag,
};

use crate::client::{BffClient, enc};
use crate::dto::album::AlbumCardDto;
use crate::dto::artist::ArtistCardDto;
use crate::dto::discover::{DiscoverSummaryDto, SpotlightItemDto, SpotlightResponseDto, TagDto};
use crate::dto::envelope::ItemsEnvelope;
use crate::error::BffError;

impl BffClient {
    pub async fn discover_summary(&self) -> Result<DiscoverSummary, BffError> {
        let dto: DiscoverSummaryDto = self.get_json("/discover/summary").await?;
        Ok(dto.into_domain())
    }

    /// Каталог артистов — курсорная пагинация (`cursor` ↔ `next_cursor`).
    /// `sort` (popular/trending/...), `tag` и `q` опциональны.
    pub async fn discover_artists(
        &self,
        limit: u32,
        cursor: Option<&str>,
        sort: Option<&str>,
        tag: Option<&str>,
        q: Option<&str>,
    ) -> Result<CursorPage<ArtistCard>, BffError> {
        let mut path = format!("/discover/artists?limit={limit}");
        push_opt(&mut path, "cursor", cursor);
        push_opt(&mut path, "sort", sort);
        push_opt(&mut path, "tag", tag);
        push_opt(&mut path, "q", q);
        let env: ItemsEnvelope<ArtistCardDto> = self.get_json(&path).await?;
        Ok(CursorPage::new(
            env.items.into_iter().map(ArtistCardDto::into_domain).collect(),
            env.next_cursor,
        ))
    }

    /// Каталог альбомов — курсорная пагинация. `sort`, `kind` (album/single/ep)
    /// и `q` опциональны.
    pub async fn discover_albums(
        &self,
        limit: u32,
        cursor: Option<&str>,
        sort: Option<&str>,
        kind: Option<&str>,
        q: Option<&str>,
    ) -> Result<CursorPage<AlbumCard>, BffError> {
        let mut path = format!("/discover/albums?limit={limit}");
        push_opt(&mut path, "cursor", cursor);
        push_opt(&mut path, "sort", sort);
        push_opt(&mut path, "kind", kind);
        push_opt(&mut path, "q", q);
        let env: ItemsEnvelope<AlbumCardDto> = self.get_json(&path).await?;
        Ok(CursorPage::new(
            env.items.into_iter().map(AlbumCardDto::into_domain).collect(),
            env.next_cursor,
        ))
    }

    /// Альбомы, сгруппированные по годам (`/discover/albums/by-year`).
    pub async fn discover_albums_by_year(
        &self,
        years: u32,
        per_year: u32,
        kind: Option<&str>,
    ) -> Result<Vec<AlbumYearBucket>, BffError> {
        let mut path = format!("/discover/albums/by-year?years={years}&per_year={per_year}");
        push_opt(&mut path, "kind", kind);
        let dto: YearBucketsDto = self.get_json(&path).await?;
        Ok(dto.into_domain())
    }

    /// Случайный артефакт для «surprise me» (`/discover/random?type=`).
    /// `Ok(None)` если каталог пуст (404).
    pub async fn discover_random(&self, kind: Option<&str>) -> Result<Option<String>, BffError> {
        let mut path = String::from("/discover/random");
        if let Some(kind) = kind.filter(|v| !v.is_empty()) {
            path.push_str("?type=");
            path.push_str(&enc(kind));
        }
        let dto: Option<RandomDto> = self.get_optional(&path).await?;
        Ok(dto.map(|d| d.id))
    }

    pub async fn discover_tags(&self) -> Result<Vec<Tag>, BffError> {
        let env: ItemsEnvelope<TagDto> = self.get_json("/discover/tags").await?;
        Ok(env.items.into_iter().map(TagDto::into_domain).collect())
    }

    /// «В центре внимания» (`/discover/spotlight`) — курируемые карточки
    /// (артист|альбом), без пагинации. `limit` опционален.
    pub async fn discover_spotlight(
        &self,
        limit: Option<u32>,
    ) -> Result<Vec<SpotlightItem>, BffError> {
        let mut path = String::from("/discover/spotlight");
        if let Some(limit) = limit.filter(|v| *v > 0) {
            path.push_str("?limit=");
            path.push_str(&limit.to_string());
        }
        let dto: SpotlightResponseDto = self.get_json(&path).await?;
        Ok(dto.items.into_iter().map(SpotlightItemDto::into_domain).collect())
    }
}

fn push_opt(path: &mut String, key: &str, value: Option<&str>) {
    if let Some(value) = value.filter(|v| !v.is_empty()) {
        path.push('&');
        path.push_str(key);
        path.push('=');
        path.push_str(&enc(value));
    }
}

#[derive(serde::Deserialize)]
struct RandomDto {
    id: String,
}

#[derive(serde::Deserialize)]
struct YearBucketsDto {
    #[serde(default)]
    buckets: Vec<YearBucketDto>,
}

#[derive(serde::Deserialize)]
struct YearBucketDto {
    #[serde(default)]
    year: i32,
    #[serde(default)]
    items: Vec<AlbumCardDto>,
}

impl YearBucketsDto {
    fn into_domain(self) -> Vec<AlbumYearBucket> {
        self.buckets
            .into_iter()
            .map(|b| AlbumYearBucket {
                year: b.year,
                items: b.items.into_iter().map(AlbumCardDto::into_domain).collect(),
            })
            .collect()
    }
}
